src/lib.rs

```diff
@@ -19,78 +19,37 @@ fn _byte_pair_merge(
     ranks: &HashMap<Vec<u8>, Rank>,
     piece: &[u8],
 ) -> Vec<(usize, Rank)> {
-    // This is a vector of (start, rank).
-    // The rank is of the byte pair starting at position start.
-    // The rank of the last item in the vector is not a valid value.
-    let mut parts: Vec<(usize, Rank)> = (0..piece.len() + 1).map(|i| (i, Rank::MAX)).collect();
-
-    let get_rank = {
-        #[inline(always)]
-        |parts: &Vec<(usize, Rank)>, start_idx: usize, skip: usize| {
-            if (start_idx + skip + 2) < parts.len() {
-                ranks
-                    .get(&piece[parts[start_idx].0..parts[start_idx + skip + 2].0])
-                    .copied()
-            } else {
-                None
-            }
-        }
+    let get_rank = |parts: &Vec<(usize, _)>, start_idx: usize, end_idx: usize| {
+        *parts.get(end_idx)
+            .map(|e| parts.get(start_idx).unwrap().0..e.0)
+            .and_then(|r| piece.get(r))
+            .filter(|p| p.len() < piece.len())
```

Comment:
minor: since we've filtered out single-tokens, if the range returns the whole piece we just default to `&Rank::MAX)`, saving a loopup